### ISSUE TITLE
NaN Loss bug and a step-by-step code analysis

### ISSUE THREAD

**Issue by tdrvlad** [tensorflow/addons#2447]
Following on problems I had while training with Triplet Loss, I went through the backend code and gathered some personal / visual explanations. Feel free to use my resources.

https://tudorvladstefan.medium.com/de-mystifying-tensorflows-triplet-loss-deca00ca9479

**Comment by yacinemrabet1** [comment]
`    triplet_loss = tf.math.truediv(
        tf.math.reduce_sum(
            tf.math.maximum(tf.math.multiply(loss_mat, mask_positives), 0.0)
        ),
        num_positives,
    )`

I think they should test if num_positives is equal to zero or not.

**Comment by tdrvlad** [comment]
Hi, yes, indeed, the nan error results from there being not tough positives
in the batch (one has to take this into account, especially when dealing
with small batches) I took the time to elaborate in an article:
https://tudorvladstefan.medium.com/de-mystifying-tensorflows-triplet-loss-deca00ca9479
Cheers!
